### PR TITLE
Use value tuples in Streams

### PR DESCRIPTION
**Pull request by vasily-kirichenko**

### CODE REVIEW

**Comment by Aaronontheweb** [comment]
@vasily-kirichenko not sure why exactly, but it looks like the Mono compilation stage is having some trouble with these changes. I think we're running an older version of Mono 5.0* - should we upgrade the Docker image we're using for compilation there to see if that solves the issue? Windows compilation appears to be working just fine.

**Comment by Aaronontheweb** [comment]
.NET Core compilation on Linux works fine too.

**Comment by vasily-kirichenko** [comment]
> I think we're running an older version of Mono 5.0* - should we upgrade the Docker image we're using for compilation there to see if that solves the issue?

I've never used Mono, and I've not found any mentions of Docker in build.fsx. How to do the upgrade?

**Comment by Aaronontheweb** [comment]
> I've never used Mono, and I've not found any mentions of Docker in build.fsx. How to do the upgrade?

We run the `build.fsx` inside the Docker container built using the Dockerfile here for Mono builds: https://github.com/akkadotnet/dockerfiles

Those images eventually get pushed up here: https://hub.docker.com/r/akkadotnet/mono-base/

We use Docker for a few things in our CI system, mostly hosting the Mono environment and running Akka.Persistence integration tests, but that stuff mostly happens invisibly on our TeamCity agents and isn't exposed directly in the build scripts.

**Comment by vasily-kirichenko** [comment]
@Aaronontheweb https://github.com/akkadotnet/dockerfiles/pull/7

**Comment by marcpiechura** [comment]
@vasily-kirichenko as @Aaronontheweb mentioned in the issue, this is quite a breaking change and should therefore target the v1.4 branch. 

BTW, thanks a lot for taking this one 😄

**Comment by vasily-kirichenko** [comment]
@marcpiechura done

File: src/core/Akka.Streams/Implementation/ActorMaterializerImpl.cs [hunk 1]
```diff
@@ -123,51 +126,47 @@ public ActorMaterializerSession(ActorMaterializerImpl materializer, IModule topL
...
+                    AssignPort(stage1.Out, UntypedPublisher.FromTyped(processor));
                     materializedValues.Add(atomic, materialized);
                 }
                 //else if (atomic is TlsModule)
                 //{
                 //})
                 else if (atomic is GraphModule)
                 {
-                    var graph = (GraphModule) atomic;
+                    var graph = (GraphModule)atomic;
```
Comment (marcpiechura) [inline src/core/Akka.Streams/Implementation/ActorMaterializerImpl.cs]: Why not also use assignment in the if statement ?

File: src/core/Akka.Streams/Implementation/ActorMaterializerImpl.cs [hunk 2]
```diff
@@ -123,51 +126,47 @@ public ActorMaterializerSession(ActorMaterializerImpl materializer, IModule topL
...
                     MaterializeGraph(graph, effectiveAttributes, materializedValues);
                 }
                 else if (atomic is GraphStageModule)
                 {
-                    var stage = (GraphStageModule) atomic;
+                    var stage = (GraphStageModule)atomic;
```
Comment (marcpiechura) [inline src/core/Akka.Streams/Implementation/ActorMaterializerImpl.cs]: As above

File: src/core/Akka.Persistence.TCK/PluginSpec.cs [hunk 3]
```diff
@@ -43,19 +43,12 @@ public void Subscribe<T>(IActorRef subscriber)
             Sys.EventStream.Subscribe(subscriber, typeof (T));
         }
 
-        /// <inheritdoc/>
-        public void Dispose()
-        {
-            Dispose(true);
-            GC.SuppressFinalize(this);
-        }
-
```
[the remark this answers is not in the document]
Comment (vasily-kirichenko) [reply src/core/Akka.Persistence.TCK/PluginSpec.cs]: Yes. The same code is in the base class's Dispose method.

**Comment by vasily-kirichenko** [comment]
@marcpiechura all done.

**Comment by Aaronontheweb** [comment]
Going to pull the plug on our Mono builds, I think; been talking with @Horusiath about dropping support for it since it affects PRs like this one and our `IObservable` integration. Going to restore our .NET Core on Linux builds and use that as our canary for x-plat compatibility on POSIX operating systems.

**Comment by marcpiechura** [comment]
@Aaronontheweb any timeframe for removing mono? Would like to merge this one and rebase my work for Streams 2.5 on it

**Comment by Aaronontheweb** [comment]
I'll look at resolving the conflicts here and getting this pulled into v1.4 now that v1.4 has been updated with the latest sources.

**Comment by Aaronontheweb** [comment]
Need to give this another look for the next Akka.NET v1.4.0 beta

**Comment by Aaronontheweb** [comment]
Retargeting the `dev` branch here...

**Comment by Aaronontheweb** [comment]
The merge conflicts are a little nasty - I think I'll need to do this offline.

**Comment by Aaronontheweb** [comment]
closed via https://github.com/akkadotnet/akka.net/pull/3973

**Pull request closed without merging**